File: game-engine/src/main/java/org/luxons/sevenwonders/game/resources/Resources.java
package org.luxons.sevenwonders.game.resources;

import java.util.EnumMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.NoSuchElementException;
import java.util.Objects;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class Resources {

    private final Map<ResourceType, Integer> quantities = new EnumMap<>(ResourceType.class);

    public static Resources of(ResourceType... types) {
        Resources resources = new Resources();
        for (ResourceType type : types) {
            resources.add(type, 1);
        }
        return resources;
    }

    public void add(ResourceType type, int quantity) {
        quantities.merge(type, quantity, (x, y) -> x + y);
    }

    public void remove(ResourceType type, int quantity) {
        if (getQuantity(type) < quantity) {
            throw new NoSuchElementException(String.format("Can't remove %d resources of type %s", quantity, type));
        }
        quantities.computeIfPresent(type, (t, oldQty) -> oldQty - quantity);
    }

    public void addAll(Resources resources) {
        resources.quantities.forEach(this::add);
    }

    public int getQuantity(ResourceType type) {
        return quantities.getOrDefault(type, 0);
    }

    public List<ResourceType> asList() {
        return quantities.entrySet()
                         .stream()
                         .flatMap(e -> Stream.generate(e::getKey).limit(e.getValue()))
                         .collect(Collectors.toList());
    }

    public boolean contains(Resources resources) {
        return resources.quantities.entrySet().stream().allMatch(this::hasAtLeast);
    }

    private boolean hasAtLeast(Entry<ResourceType, Integer> quantity) {
        return quantity.getValue() <= getQuantity(quantity.getKey());
    }

    /**
     * Creates new {@link Resources} object containing these resources minus the given resources.
     *
     * @param resources
     *         the resources to subtract from these resources
     *
     * @return a new {@link Resources} object containing these resources minus the given resources.
     */
    public Resources minus(Resources resources) {
        Resources diff = new Resources();
        quantities.forEach((type, count) -> {
            int remainder = count - resources.getQuantity(type);
            diff.quantities.put(type, Math.max(0, remainder));
        });
        return diff;
    }

    public boolean isEmpty() {
        return size() == 0;
    }

    public int size() {
        return quantities.values().stream().reduce(0, Integer::sum);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        Resources resources = (Resources) o;
        return Objects.equals(quantities, resources.quantities);
    }

    @Override
    public int hashCode() {
        return Objects.hash(quantities);
    }

    @Override
    public String toString() {
        return "Resources{" + "quantities=" + quantities + '}';
    }
}

File: game-engine/src/main/kotlin/org/luxons/sevenwonders/game/resources/Resources.kt
package org.luxons.sevenwonders.game.resources

import java.util.NoSuchElementException

fun emptyResources(): Resources = MutableResources()

fun resourcesOf(singleResource: ResourceType): Resources = MutableResources(mutableMapOf(singleResource to 1))

fun resourcesOf(vararg resources: ResourceType): Resources =
    resources.fold(MutableResources()) { rs, r -> rs.add(r, 1); rs }

fun resourcesOf(resources: Iterable<ResourceType>): Resources =
    resources.fold(MutableResources()) { rs, r -> rs.add(r, 1); rs }

fun resourcesOf(vararg resources: Pair<ResourceType, Int>): Resources =
    resources.fold(MutableResources()) { rs, (type, qty) -> rs.add(type, qty); rs }

internal fun mutableResourcesOf() = MutableResources()

internal fun mutableResourcesOf(vararg resources: Pair<ResourceType, Int>) =
    resources.fold(MutableResources()) { rs, (type, qty) -> rs.add(type, qty); rs }

fun Iterable<ResourceType>.toResources(): Resources = resourcesOf(this)

fun Iterable<Resources>.merge(): Resources = fold(MutableResources()) { r1, r2 -> r1.add(r2); r1}

internal fun Resources.toMutableResources(): MutableResources {
    val resources = MutableResources()
    resources.add(this)
    return resources
}

interface Resources {

    val quantities: Map<ResourceType, Int>

    val size: Int
        get() = quantities.map { it.value }.sum()

    fun isEmpty(): Boolean = size == 0

    operator fun get(key: ResourceType): Int = quantities.getOrDefault(key, 0)

    fun containsAll(resources: Resources): Boolean = resources.quantities.all { it.value <= this[it.key] }

    operator fun plus(resources: Resources): Resources {
        val new = MutableResources()
        new.add(this)
        new.add(resources)
        return new
    }

    operator fun minus(resources: Resources): Resources {
        val diff = MutableResources()
        quantities.forEach { type, count ->
            val remainder = count - resources[type]
            diff.quantities[type] = Math.max(0, remainder)
        }
        return diff
    }

    fun toList(): List<ResourceType> = quantities.flatMap { (type, quantity) -> List(quantity) { type } }
}

data class MutableResources(
    override val quantities: MutableMap<ResourceType, Int> = mutableMapOf()
) : Resources {

    fun add(type: ResourceType, quantity: Int) {
        quantities.merge(type, quantity) { x, y -> x + y }
    }

    fun add(resources: Resources) = resources.quantities.forEach { type, quantity -> this.add(type, quantity) }

    fun remove(type: ResourceType, quantity: Int) {
        if (this[type] < quantity) {
            throw NoSuchElementException("Can't remove $quantity resources of type $type")
        }
        quantities.computeIfPresent(type) { _, oldQty -> oldQty - quantity }
        // to ensure equals() work properly
        if (quantities[type] == 0) {
            quantities.remove(type)
        }
    }
}
